pspec_run config for batch pipeline

nkern opened the pull request:
Updates to `pspec_run` given recent additions to `PSpecData` and `pyuvdata`, and also adds `utils.config_pspec_blpairs` for use by the batch pipeline.

Also fixes some leftover issues that weren't addressed in the analytic errors PR.

philbull commented on hera_pspec/pspecdata.py:
```diff
@@ -2126,6 +2138,17 @@ def pspec_run(dsets, filename, groupname=None, dset_labels=None, dset_pairs=None
+        If True, broadcast dset flags across time using fractional time_thresh.
+
+    time_thresh : float
+        Fractional flagging threshold to trigger broadcast across time if
```
Can you give a few more details about how this works, please? Is it done across all blpairs etc?

philbull commented on hera_pspec/pspecdata.py:
```diff
@@ -2171,17 +2199,32 @@ def pspec_run(dsets, filename, groupname=None, dset_labels=None, dset_pairs=None
+    if dset_pairs is None:
+        dset_pairs = list(itertools.combinations(range(Ndsets), 2))
+
+    if dset_labels is None:
```
Should we also check for duplicate dset labels? That could screw things up if it ever happens.

philbull commented on hera_pspec/pspecdata.py:
```diff
@@ -2171,17 +2199,32 @@ def pspec_run(dsets, filename, groupname=None, dset_labels=None, dset_pairs=None
+            utils.log("Loaded data in %1.1f sec." % (time.time() - t0), lvl=1, verbose=verbose)
+        except ValueError:
+            # at least one of the dset loads failed due to no data being present
+            utils.log("One of the dset loads failed due to no data overlap given the bls and pols selection", verbose=verbose)
```
Should we error out here?

philbull commented on hera_pspec/pspecdata.py:
```diff
@@ -2171,17 +2199,32 @@ def pspec_run(dsets, filename, groupname=None, dset_labels=None, dset_pairs=None
+            t0 = time.time()
+            _dsets = []
+            for i, dset in enumerate(dsets):
+                # read data
```
Since this step is IO heavy, it might be very time-intensive, so should we print progress to screen?

philbull commented on hera_pspec/pspecdata.py:
```diff
@@ -2209,15 +2252,19 @@ def pspec_run(dsets, filename, groupname=None, dset_labels=None, dset_pairs=None
+
+    # broadcast flags
+    if broadcast_dset_flags:
+        ds.broadcast_dset_flags(time_thresh=time_thresh)
```
Are the broadcast flags saved somewhere?

philbull commented on hera_pspec/utils.py:
```diff
@@ -505,3 +530,123 @@ def flatten(nested_list):
+        List of len-2 polarization tuples to use in forming cross spectra.
+        Ex: [('xx', 'xx'), ('yy', 'yy'), ...]
+
+    group_pairs : list
```
Not quite sure what a 'group' is in this context?

nkern replied:
Sure, I'll add some more text in the doc-string with examples

nkern commented on hera_pspec/pspecdata.py:
```diff
@@ -2056,10 +2057,17 @@ def pspec_run(dsets, filename, groupname=None, dset_labels=None, dset_pairs=None
         Default is to form all N_choose_2 pairs from input dsets.
 
+    psname_ext : string
+        A string extension for the psname in the container object.
```
Sure

nkern commented:
@philbull okay addressed your comments, merged in `master` and fixed some of the issues leftover from the Analytic Errors PR.

I would like to merge this _after_ the `stats_array` PR is merged in, so let's wait for the time being until AEW and DR get that PR cleaned up.

nkern commented:
This PR is being merged into the __power spectrum pipeline v1__ PR (#151 ).
